crates/bevy_reflect/src/set.rs

```diff
@@ -0,0 +1,506 @@
+use std::any::{Any, TypeId};
+use std::fmt::{Debug, Formatter};
+
+use bevy_reflect_derive::impl_type_path;
+use bevy_utils::{Entry, HashMap};
+
+use crate::{
+    self as bevy_reflect, Reflect, ReflectKind, ReflectMut, ReflectOwned, ReflectRef, TypeInfo,
+    TypePath, TypePathTable,
+};
+
+/// A trait used to power [set-like] operations via [reflection].
+///
+/// Set contain zero or more entries of a value, and correspond to types like [`HashSet`]. The
+/// order of these entries is not guaranteed by this trait.
+///
+/// # Hashing
+///
+/// All keys are expected to return a valid hash value from [`Reflect::reflect_hash`].
+/// If using the [`#[derive(Reflect)]`](derive@crate::Reflect) macro, this can be done by adding `#[reflect(Hash)]`
+/// to the entire struct or enum.
+/// This is true even for manual implementors who do not use the hashed value,
+/// as it is still relied on by [`DynamicSet`].
+///
+/// # Example
+///
+/// ```
+/// use bevy_reflect::{Reflect, Set};
+/// use bevy_utils::HashSet;
+///
+///
+/// let foo: &mut dyn Set = &mut HashSet::<u32>::new();
+/// foo.insert_boxed(Box::new(123_u32));
+/// assert_eq!(foo.len(), 1);
+///
+/// let field: &dyn Reflect = foo.get(&123_u32).unwrap();
+/// assert_eq!(field.downcast_ref::<u32>(), Some(&123_u32));
+/// ```
+///
+/// [set-like]: https://doc.rust-lang.org/stable/std/collections/struct.HashSet.html
+/// [reflection]: crate
+pub trait Set: Reflect {
```

Comment:
I agree with your assessment, it would make sense to add a `contains` method to this trait.

I don't know about adding union/intersection methods _yet_. Maybe that can come in a future PR where we can assess if most set-like types would be able to support those operations (I imagine they would but it might be good to have a dedicated discussion around it).